src/system.c

```diff
@@ -1328,4 +1334,20 @@ void _flush_cache(uint8_t* addr, unsigned long bytes) {
     inst_cache_hit_invalidate(addr, bytes);
 }
 
+/**
+ * @brief Implement underlying function for assert()
+ *
+ * Implementation of the function called when an assert fails. By default,
+ * we just abort execution, but this will be overriden at startup with
```

Comment:
libdragon is compiled in two libraries `libdragonsys.a` and `libdragon.a`. `libdragonsys` contains the "OS backend" of newlib (`system.c` mostly), while `libdragon.a` contains everything else.

This is because historically C linkers have not been able to solve circular dependencies of symbols among libraries and required command line invocations to list libraries in strict dependency order. So the order we expect is `game` -> `libdragon` -> `newlib` -> `libdragonsys.a` (or even `game` -> `newlib` -> `libdragonsys.a`). This is why all examples have the following library list: `-ldragon -lc -lm -ldragonsys` in that order.

So we need to implement `__assert_func` which is called by newlib, so it must stay in `libdragonsys.a`. At the same time, we want the "real" assert code to be in debug.c and it must be able to call the console library etc., so it must stay in `libdragon`. To break this dependency, we add the function pointer indirection.

Notice that since we are using a modern linker, we could ask the linker to solve circular dependencies; for reasons that elude me (maybe performance), that's not on by default, so if we start requiring that, it would break all linker scripts in the world, with strange errors about missing symbols that nobody would know how to debug or fix. 

So at the end of the day, it's better to keep it like it is right now, and respecting the split between low-level system code and higher level functionalities.

Reply:
I now realized why, thanks. Maybe link time optimizations can handle that and remove the additional call